subsys/bluetooth/mesh/dfu_slot.c

```diff
@@ -147,145 +139,176 @@ bt_mesh_dfu_slot_add(size_t size, const uint8_t *fwid,
+	slot->slot.fwid_len = 0;
+	slot->slot.metadata_len = 0;
+	slot->slot.size = 0;
+	slot->idx = ++slot_index;
+
+	LOG_DBG("Reserved slot #%u", slot - &slots[0]);
 
-	LOG_DBG("Added slot #%u: %s", slot - &slots[0],
-		bt_hex(slot->slot.fwid, slot->slot.fwid_len));
 	return &slot->slot;
 }
 
-int bt_mesh_dfu_slot_valid_set(const struct bt_mesh_dfu_slot *slot, bool valid)
+int bt_mesh_dfu_slot_info_set(struct bt_mesh_dfu_slot *dfu_slot, size_t size,
+			      const uint8_t *metadata, size_t metadata_len)
 {
-	uint16_t idx;
-	bool prev;
-	int err;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (metadata_len > CONFIG_BT_MESH_DFU_METADATA_MAXLEN) {
+		return -EFBIG;
+	}
+
+	if (slot->idx == 0 || is_slot_committed(slot)) {
+		return -EINVAL;
+	}
+
+	slot->slot.size = size;
+	slot->slot.metadata_len = metadata_len;
+	memcpy(slot->slot.metadata, metadata, metadata_len);
+	return 0;
+}
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return -ENOENT;
+int bt_mesh_dfu_slot_fwid_set(struct bt_mesh_dfu_slot *dfu_slot,
+			      const uint8_t *fwid, size_t fwid_len)
+{
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
+
+	if (fwid_len > CONFIG_BT_MESH_DFU_FWID_MAXLEN) {
+		return -EFBIG;
+	}
+
+	if (slot->idx == 0 || is_slot_committed(slot)) {
+		return -EINVAL;
 	}
 
-	idx = slot_idx(slot);
+	for (int i = 0; i < ARRAY_SIZE(slots); i++) {
+		if (slots[i].idx != 0 &&
+		    slot_eq(&slots[i].slot, fwid, fwid_len)) {
+			return -EALREADY;
+		}
+	}
+
+	slot->slot.fwid_len = fwid_len;
+	memcpy(slot->slot.fwid, fwid, fwid_len);
+	return 0;
+}
 
-	LOG_DBG("%u: %u", idx, valid);
+int bt_mesh_dfu_slot_commit(struct bt_mesh_dfu_slot *dfu_slot)
+{
+	int err;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (valid) {
-		prev = atomic_test_and_set_bit(valid_slots, idx);
-	} else {
-		prev = atomic_test_and_clear_bit(valid_slots, idx);
+	if (slot->idx == 0 ||
+	    slot->slot.fwid_len == 0 ||
+	    slot->slot.size == 0 ||
+	    is_slot_committed(slot)) {
+		return -EINVAL;
 	}
 
-	if (valid == prev) {
-		return 0;
+	if (bt_mesh_dfu_slot_get(slot->slot.fwid, slot->slot.fwid_len, NULL)
+	    >= 0) {
+		return -EALREADY;
 	}
 
-	err = valid_slots_store();
+	err = slot_store(slot);
 	if (err) {
-		LOG_WRN("Storage failed. err: %d", err);
-		atomic_set_bit_to(valid_slots, idx, prev);
+		LOG_WRN("Store failed (err: %d)", err);
+		return err;
 	}
 
-	return err;
+	sys_slist_append(&list, &slot->n);
+
+	LOG_DBG("Stored slot #%u: %s", ARRAY_INDEX(slots, slot),
+		bt_hex(slot->slot.fwid, slot->slot.fwid_len));
+	return 0;
 }
 
-bool bt_mesh_dfu_slot_is_valid(const struct bt_mesh_dfu_slot *slot)
+int bt_mesh_dfu_slot_release(const struct bt_mesh_dfu_slot *dfu_slot)
 {
-	uint16_t idx;
+	struct slot *slot = CONTAINER_OF(dfu_slot, struct slot, slot);
 
-	if (!SLOT_IN_ARRAY(slot) || !slot_in_use(slot)) {
-		return false;
+	if (slot->idx == 0 || is_slot_committed(slot)) {
+		return -EINVAL;
```

Comment:
We don't really check return code of `bt_mesh_dfu_slot_release`. If we don't need it, perhaps `bt_mesh_dfu_slot_release` can return `void`?